src_c/_camera.c

```diff
@@ -160,41 +160,46 @@ list_cameras(PyObject *self, PyObject *arg)
 #else
     WCHAR **devices;
 #endif
-    int num_devices, i;
-
-    num_devices = 0;
-    ret_list = NULL;
-    ret_list = PyList_New(0);
-    if (!ret_list)
-        return NULL;
+    int j, i = 0, num_devices = 0;
 
+    /* TODO for future PRs: errors in these functions are being ignored as
+     * of now, and an empty list is being returned by this function */
 #if defined(__unix__)
     devices = v4l2_list_cameras(&num_devices);
 #elif defined(PYGAME_WINDOWS_CAMERA)
     devices = windows_list_cameras(&num_devices);
 #endif
+
+    ret_list = PyList_New(num_devices);
+    if (!ret_list) {
+        goto error;
+    }
+
     for (i = 0; i < num_devices; i++) {
 #if defined(PYGAME_WINDOWS_CAMERA)
         string = PyUnicode_FromWideChar(devices[i], -1);
 #else
         string = PyUnicode_FromString(devices[i]);
 #endif
-        if (0 != PyList_Append(ret_list, string)) {
-            /* Append failed; clean up and return */
-            Py_DECREF(ret_list);
-            Py_DECREF(string);
-            for (; i < num_devices; i++) {
-                free(devices[i]);
-            }
-            free(devices);
-            return NULL; /* Exception already set. */
+        if (!string) {
+            goto error;
         }
-        Py_DECREF(string);
+        /* steals reference to 'string' */
+        PyList_SET_ITEM(ret_list, i, string);
```

Comment:
Should we use the function form instead to stay closer to the stable ABI?

Reply:
Well... A lot of pygame code uses non-stable ABI code ATM... porting to stable ABI will be hard anyways, because even basic and important structures like `PyTypeObject` are not a part of the limited API. So if we ever need to support it, it would basically mean rewriting most of pygame, either implementing all pygame classes in python with calls to internal pygame functions written in C, or maybe dynamic classes in C code, which I am not really sure, will definitely be harder than a python rewrite. This could also be handled if we ever do a `HPy` port.

TL;DR stable ABI is still very far from now, so better to not worry about it for now

Reply:
PyTypeObject is part of the stable ABI.

I'm not pushing the stable ABI hard because we already have good infrastructure to handle wheel generation, and the stable ABI can be a bit slower in places.

Up to you to resolve.

Reply:
`PyTypeObject` is definitely not a part of the stable ABI. This is something I got stuck with when I was trying to port a small C extension to the stable ABI.
According to the [docs](https://docs.python.org/3/c-api/typeobj.html#static-types)
> Also, since [PyTypeObject](https://docs.python.org/3/c-api/type.html#c.PyTypeObject) is not part of the [stable ABI](https://docs.python.org/3/c-api/stable.html#stable), any extension modules using static types must be compiled for a specific Python minor version.

So for the same reason, I will leave this function in macro form for now, because it can be a tad bit faster and the code is also simpler without unnecessary error checking